### PR TITLE
dropck_outlives check whether generator witness needs_drop

### LINKED ISSUE #116242: Code no longer compiles after -Zdrop-tracking-mir was enabled by default
Some code that used to compile no longer does after #107421. We don't think was ever *supposed* to compile. Don't have an MCVE yet.

Snippet:

```rust
let selecting_fut = selecting
    .do_selecting(
        &client_config,
        &test_socket_provider,
        &mut rng,
        &time,
        &mut stop_receiver,
    )
    .fuse();

let time = &time;

let wait_fut = async {
    // Wait some arbitrary amount of time to ensure `do_selecting` is waiting on a reply.
    // Note that this is fake time, not 30 actual seconds.
    time.wait_until(std::time::Duration::from_secs(30)).await;
}
.fuse();

pin_mut!(selecting_fut, wait_fut);

let main_future = async {
    select! {
        _ = selecting_fut => unreachable!("should keep retransmitting DHCPDISCOVER forever"),
        () = wait_fut => (),
    }
};
pin_mut!(main_future);

run_with_accelerated_time(&mut executor, time, &mut main_future);

stop_sender.unbounded_send(()).expect("sending stop signal should succeed");

let selecting_result = selecting_fut.now_or_never().expect(
    "selecting_fut should complete after single poll after stop signal has been sent",
);

assert_matches!(selecting_result, Ok(SelectingOutcome::GracefulShutdown));
```

Full code is here: https://cs.opensource.google/fuchsia/fuchsia/+/main:src/connectivity/network/dhcpv4/client/core/src/client.rs;l=1521-1576;drc=4a7d2eb6793e61c0021f8dd7dab35804590d36b6

I expected to see this happen: *compiles as before*

Instead, this happened:

```
error[E0505]: cannot move out of `selecting_fut` because it is borrowed
    --> ../../src/connectivity/network/dhcpv4/client/core/src/client.rs:1571:32
     |
1559 |           let main_future = async {
     |  ___________________________-
1560 | |             select! {
1561 | |                 _ = selecting_fut => unreachable!("should keep retransmitting DHCPDISCOVER forever"),
     | |                     ------------- borrow occurs due to use in generator
1562 | |                 () = wait_fut => (),
1563 | |             }
1564 | |         };
     | |_________- borrow of `selecting_fut` occurs here
...
1571 |           let selecting_result = selecting_fut.now_or_never().expect(
     |                                  ^^^^^^^^^^^^^ move out of `selecting_fut` occurs here
...
1576 |       }
     |       - borrow might be used here, when `main_future` is dropped and runs the destructor for generator

error: aborting due to previous error

For more information about this error, try `rustc --explain E0505`.
```

### Version it worked on

It most recently worked on: Nightly 2023-09-22

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

### PR DESCRIPTION
**Pull request by lcnr**
see https://rust-lang.zulipchat.com/#narrow/stream/326866-t-types.2Fnominated/topic/.23116242.3A.20Code.20no.20longer.20compiles.20after.20-Zdrop-tracking-mir.20.E2.80.A6/near/398311627 for an explanation.

Fixes #116242 (or well, the repro by @jamuraa in https://github.com/rust-lang/rust/issues/116242#issuecomment-1739802047). I did not add a regression test as it depends on other crates. We do have 1 test going from fail to pass, showing the intended behavior.

r? types

### CODE REVIEW

File: compiler/rustc_ty_utils/src/needs_drop.rs [hunk 1]
```diff
@@ -133,8 +137,31 @@ where
...
+                    //
+                    // HACK: Because we erase regions contained in the generator witness, we
+                    // have to conservatively assume that every region captured by the
+                    // generator has to be live when dropped. This results in a lot of
+                    // undesirable borrowck errors. During borrowck, we call `needs_drop`
+                    // for the generator witness and check whether any of the contained types
+                    // need to be dropped, and only require the captured types to be live
+                    // if they do.
+                    ty::Coroutine(_, args, _) => {
+                        if self.reveal_coroutine_witnesses {
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/needs_drop.rs:149]: this is necessary to handle nested futures :thinking: should potentially add a test for it, but I don't feel motivated enough to do so :sweat_smile:

**Comment by lcnr** [comment]
want to run both crater and perf. Worried that I messed up and we get some weird cycles because of this, I am fairly confident that this is alright.

**Comment by cjgillot** [comment]
Thanks for figuring this out @lcnr!

**Comment by lcnr** [comment]
@rust-timer queue

@craterbot check

**Comment by lcnr** [comment]
assuming that the crater run and perf is acceptable, and to make sure we backport in time:

To quickly summarize again: *changes from this PR in **bold***
- dropping a coroutine witness drops the values stored in its state machine: the content of its `CoroutineWitness`
- computing the `CoroutineWitness` relies on `mir_built`
- `mir_built` relies on `fn needs_drop`, we conservatively assume that all generators have to be dropped
- we do not know the exact lifetimes of the values stored in the witness (as they may be self-referential), so for a generator to be *drop live* during `mir_borrowck` we require anything that could theoretically be accessed by the coroutine to be live.
- however, during `mir_borrowck` we can access the `CoroutineWitness`, so **we only require everything to be live if the coroutine actually contains any field which requires drop. This recursively looks into nested coroutines.**

I believe that this does not result in query cycles as the only way `needs_drop` can access nested coroutines if they are from the same body. Coroutines from other bodies are hidden by an opaque type, which we should not be able to reveal.

@rfcbot fcp merge

**Comment by lcnr** [comment]
@craterbot p=1

**Comment by lqd** [comment]
@rust-timer build 2cf1c2a614d6ae3bd50a8df20aeac2f9eceeea87

File: compiler/rustc_ty_utils/src/needs_drop.rs [hunk 2]
```diff
@@ -66,6 +66,9 @@ fn has_significant_drop_raw<'tcx>(
 struct NeedsDropTypes<'tcx, F> {
     tcx: TyCtxt<'tcx>,
     param_env: ty::ParamEnv<'tcx>,
+    // Whether to reveal coroutine witnesses, this is set
+    // to `false` unless we compute `needs_drop` for a generator witness.
```
Comment (lqd) [inline compiler/rustc_ty_utils/src/needs_drop.rs:70]: Nit: there are a couple remaining mentions of "generators"

```suggestion
    // to `false` unless we compute `needs_drop` for a coroutine witness.
```

**Comment by lqd** [comment]
> I did not add a regression test as it depends on other crates

https://github.com/rust-lang/rust/issues/116242#issuecomment-1780891997 should help.

**Comment by wesleywiser** [comment]
Going to go ahead and mark this as

@bors p=5

because this fixes a P-critical stable-to-beta regression that we want to land expediently once there is consensus.

**Comment by apiraino** [comment]
T-compiler thinks this could be worth a backport ([Zulip notes](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-10-26/near/398697959)) but let's wait the crater run results first.

**Comment by nikomatsakis** [comment]
@lcnr 

> we only require everything to be live if the coroutine actually contains any field which requires drop

can you say a bit more about this -- this makes it sound 'binary', but my expectation was that we would treat the 'witness' the same way we do any struct that does not implement `Drop`. i.e., it does not require lifetimes to be drop-live unless they are found within some field whose type does implement `Drop`.

**Comment by lcnr** [comment]
> can you say a bit more about this -- this makes it sound 'binary', but my expectation was that we would treat the 'witness' the same way we do any struct that does not implement `Drop`. i.e., it does not require lifetimes to be drop-live unless they are found within some field whose type does implement `Drop`.

We don't look into the witness itself because lifetimes in the witness are all replaced with unconstrained bound variables.

So if we get something like `&mut Option<NeedsUseLive<'a>>` as an upvar to the coroutine, the coroutine could take the inner value and contain `NeedsUseLive<'a>` in its witness.

Because we erase **all** lifetimes found in the coroutine, the witness only contains `for/exists<'a> NeedsUseLive<'a>` and we can't really reason about the original lifetime for `'a`. I originally went with the approach you described, but that ended up being unsound :grin:

**Comment by Mark-Simulacrum** [comment]
@craterbot retry-report

**Comment by lqd** [comment]
Actually, looking at the queue, because this is testing a P-critical issue fix: @craterbot p=2

**Comment by compiler-errors** [comment]
Glad to see no crater regressions

@rustbot reviewed

**Comment by lcnr** [comment]
> (this should also fix the issue we already have for phantomdata in nested types for coroutine, going to write a test)

Yeah, same root cause as https://github.com/rust-lang/rust/issues/110288#issuecomment-1517872777
```rust
use std::future;

fn main() {
    let mut recv = future::ready(());
    let _combined_fut = async {
        let _ = || read(&mut recv);
    };
    
    let _uwu = (String::new(), _combined_fut);
    // With this line this example failed on master but will now
    // pass with this PR. The root cause is the same as 

    drop(recv);
}

fn read<F: future::Future>(_: &mut F) -> F::Output {
    todo!()
}
```

**Comment by wesleywiser** [comment]
Do we feel it's necessary to wait the full 10 days for the final comment period prior to merging? Doing so would give us less than one week of testing on nightly/beta before this change hits stable. 

If there are no objections, I'd suggest we go ahead and merge this now.

**Comment by lcnr** [comment]
I still have to rebase + deal with the reviews and originally wanted to give @nikomatsakis some time to look at this. But I think it's fine to go ahead and merge this PR today 🤔

**Comment by lcnr** [comment]
now rdy for merge after ci is green

File: tests/ui/dropck/coroutine-liveness-1.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// check-pass
```
Comment (lqd) [inline tests/ui/dropck/coroutine-liveness-1.rs:1]:
```suggestion
// edition:2021
// check-pass
```

File: tests/ui/dropck/coroutine-liveness-2.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// check-pass
```
Comment (lqd) [inline tests/ui/dropck/coroutine-liveness-2.rs:1]:
```suggestion
// edition:2021
// check-pass
```

**Comment by wesleywiser** [comment]
Accepting for beta backport [as discussed](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-11-08/near/400976394) during a previous compiler team triage meeting.

**Pull request merged**